src/combinators/mod.rs

```diff
@@ -964,6 +1029,36 @@ impl<'a> Skip for io::Cursor<&'a mut [u8]> {
     }
 }
 
+impl BackToTheBuffer for &mut [u8] {
+    fn reserve_write_use<Tmp, Gen: Fn(Self) -> GenResult<(Self, Tmp)>, Before: Fn(Self, Tmp) -> GenResult<Self>>(self, reserved: usize, gen: &Gen, before: &Before) -> GenResult<Self> {
+        let (res, buf) = self.split_at_mut(reserved);
+        let (buf, tmp) = gen(buf)?;
+        let res = before(res, tmp)?;
+        if !res.is_empty() {
+            return Err(GenError::BufferTooBig(res.len()));
+        }
+        Ok(buf)
+    }
+}
+
+#[cfg(feature = "std")]
+impl BackToTheBuffer for Vec<u8> {
+    fn reserve_write_use<Tmp, Gen: Fn(Self) -> GenResult<(Self, Tmp)>, Before: Fn(Self, Tmp) -> GenResult<Self>>(mut self, reserved: usize, gen: &Gen, before: &Before) -> GenResult<Self> {
+        let start_len = self.len();
+        self.reserve(reserved);
+        unsafe { self.set_len(start_len + reserved) };
+        let (mut vec, tmp) = gen(self)?;
+        let tmp_vec = before(Vec::new(), tmp)?;
+        let tmp_written = tmp_vec.len();
+        if tmp_written != reserved {
+            unsafe { self.set_len(start_len) };
+            return Err(GenError::BufferTooBig(reserved - tmp_written));
```

Comment:
The problem with not panicking here now is that depending on the type `W` we either ignore that less was written but the result of the writes are all in place (byte slices), or we reset the `W` so it's in the state from before any write (`Vec` impl here).

That means that the only thing generic combinators can do here is to propagate the error further upwards, and generally this error should be interpreted as "the whole serialization failed, whatever data you have here is useless"

Reply:
Well, until we have a way to deal with #22 I think than considering any failure as a whole is the best thing to do anyways

Reply:
Ok with me then, but should probably be documented somewhere :)

Reply:
Changed it back to work like any other impl